src/_posts/2019-12-31-test-driving-a-cdn.md

```diff
@@ -0,0 +1,231 @@
+
+It's possible to conditionally add headers (so that the origin, while public will only respond if sent the correct header). This allows developers to check that the origin is healthy - which will makes isolating problems much easier.
+
+You can also change the response so that details about the application are masked. It is amazing how many sites advertise the exact version of a webserver that they are using. 
+This can be seen simply using curl: `curl -v bbc.co.uk`
+
+You can use the CDN to alter the behaviour of a hosted website without changing the hosted site itself. This can allow you to apply a quick fix for a problem while a real solution is being prepared. For example you can put a holding page up for a specific url.
+
+It's advisable not to tell a CDN to cache forever as any mistakes may reside in a browser indefinately - this can be a problem if you are caching a javacript library and accidently cache a broken version.
+
+I would also recommend configuring your CDN programatically so that it is testable and repeatable. Be very careful of having the CDN do too much work as you will encounter odd edge conditions such as having an error page cached.
+
+Make sure that you know how to purge a specific page. The site that I worked on added this to a chatbot in slack. It was able to purge a page by sending it a slack message - this was ideal when you had to support a problem and were away from your machine.
+
+It is possible to use AWS S3 as a cache but it's advisable to add a CDN in front of it so that you can control the returned headers.
+
+## Learnings while preparing the sample.
+
+Alpine docker images are reasonably fast to start and download compared to the full fat versions. These are the very small docker images
+
+The tests need to wait on the varnish server being started. Normally I would have used the wait-for-it.sh script that allows waiting for a given service to be completed. However since I am using an alpine docker container we don't have bash available. This means that I have used the golang `waitforit` utility. 
+
+Also note the various network options used in the docker-compose. You can use an alias of another docker container name as a local dns entry if it is listed in the depends_on section. This is even accessable to the tests run inside the container.
+
+## What about other CDNs
+
+Now if you really want to test a real life CDN there is this project: `https://github.com/Mahoney/wiremock-heroku`
+
+This allows you to deploy wiremock into Heroku.
+
+If you configure your CDN to serve this Heroku app as the origin then you can completely test the behaviour of the CDN.
+Wiremock gives you a programmable web server that gives you an API that allows the response to change, so it's possible to have a page return a fix value once then start returning errors. This provides the ability for you to test the entire of the HTTP spec should you wish to do so, but this is beyond the scope of this article.
```

Comment:
"the entire of the HTTP spec" -> "the entirety of the HTTP spec"

Reply:
Applied